### PR TITLE
Custom Fluid-fluid interaction groups

### PR DESCRIPTION
**Pull request by ThierryBerger**
- Follow up to https://github.com/dimforge/salva/pull/62

in the following screenshot, blue and green fluid layers are not colliding:

Screenshot from 2024-11-19 14-41-27

### CODE REVIEW

**Comment by ThierryBerger** [comment]
Tagging @Ughuuu as you expressed interest in this.

**Comment by Ughuuu** [comment]
Woah! Will try to integrate into godot-rapier sometime also!

File: src/geometry/contacts.rs [hunk 1]
```diff
@@ -273,8 +273,10 @@ fn compute_contacts_for_pair_of_cells(
                         HGridEntry::BoundaryParticle(boundary_j, particle_j) => {
                             let bi = &boundaries[*boundary_i];
                             let bj = &boundaries[*boundary_j];
-                            if !bi.fluid_interaction.test(bj.fluid_interaction) {
-                                continue;
+                            if *boundary_i != *boundary_j {
+                                if !bi.interaction_groups.test(bj.interaction_groups) {
+                                    continue;
+                                }
                             }
```
Comment (sebcrozet) [inline src/geometry/contacts.rs]:
```suggestion
                            if *boundary_i != *boundary_j && !bi.interaction_groups.test(bj.interaction_groups) {
                                continue;
                            }
```

File: src/geometry/contacts.rs [hunk 2]
```diff
@@ -343,13 +345,24 @@ fn compute_contacts_for_pair_of_cells(
...
+                        if *fluid_i != fluid_j {
+                            let fluid_i_data = &fluids[*fluid_i];
+                            let fluid_j_data = &fluids[fluid_j];
+
+                            if !fluid_i_data
+                                .interaction_groups
+                                .test(fluid_j_data.interaction_groups)
+                            {
+                                continue;
+                            }
```
Comment (sebcrozet) [inline src/geometry/contacts.rs]: Not very important. Just aesthetic preferences:
```suggestion
                            let groups_i = &fluids[*fluid_i].interaction_groups;
                            let groups_j = &fluids[fluid_j].interaction_groups;

                            if !groups_i.test(groups_j) {
                                continue;
                            }
```

**Review by sebcrozet — APPROVED** [review]
Thanks!

**Pull request merged**